kong/tools/queue.lua

```diff
@@ -0,0 +1,430 @@
+    },
+    batch_max_size = {
+      type = "number",
+      default = 1,
+      description = "maximum number of entries to be given to the handler as a batch"
+    },
+    max_delay = {
+      type = "number",
+      default = 1,
+      description = "maximum number of (fractional) seconds to elapse after the first entry was queued before the queue starts calling the handler",
+    },
+    capacity = {
+      type = "number",
+      default = 10000,
+      description = "maximum number of entries that can be waiting on the queue",
+    },
+    string_capacity = {
+      type = "number",
+      default = nil,
+      description = "maximum number of bytes that can be waiting on a queue, requires string content",
+    },
+    max_retry_time = {
+      type = "number",
+      default = 60,
+      description = "time in seconds before the queue gives up calling a failed handler for a batch",
+    },
+    max_retry_delay = {
+      type = "number",
+      default = 60,
+      description = "maximum time in seconds between retries, caps exponential backoff"
+    },
+    poll_time = {
+      type = "number",
+      default = 1,
+      description = "time in seconds between polls for worker shutdown",
+    },
+    max_idle_time = {
+      type = "number",
+      default = 60,
+      description = "time in seconds before an idle queue is deleted",
+    },
+  },
+  configuration_schema = {}
+}
+
+for name, schema in pairs(Queue.configuration_fields) do
+  -- can't use schema directly because Josh's `description` has not yet been implemented.
+  table.insert(Queue.configuration_schema, { [name] = { type = schema.type, default = schema.default } })
+end
+
+
+local Queue_mt = {
+  __index = Queue
+}
+
+
+---------
+-- Queue
+---------
+
+
+local function make_queue_name(plugin_name, queue_name)
+  return plugin_name .. "." .. queue_name
+end
+
+local queues = {}
+
+
+function Queue.exists(plugin_name, queue_name)
+  return queues[make_queue_name(plugin_name, queue_name)] and true or false
+end
+
+-------------------------------------------------------------------------------
+-- Initialize a queue with background retryable processing
+-- @param process function, invoked to process every payload generated
+-- @param opts table, requires `name`, optionally includes `retry_count`, `max_delay` and `batch_max_size`
+-- @return table: a Queue object.
+function Queue.get(plugin_name, handler, opts)
+
+  assert(type(plugin_name) == "string",
+    "arg #1 (plugin_name) must be a string")
+  assert(type(handler) == "function",
+    "arg #2 (handler) must be a function")
+  assert(type(opts) == "table",
+    "arg #3 (opts) must be a table")
+  assert(type(opts.name) == "string",
+    "opts.name must be a string")
+
+  local queue_name = make_queue_name(plugin_name, opts.name)
+  local queue = queues[queue_name]
+  if queue then
+    queue:log(DEBUG, "queue exists")
+    for name, _ in pairs(Queue.configuration_fields) do
+      if queue[name] ~= opts[name] then
```

Comment:
Please don't put too much effort into refactoring this, as it will completely be replaced in 3.2.

Reply:
I'm not sure I understand.  How would `opts[name]` ever be `nil`, given the type check right above?  Splitting into separate functions would just duplicate code on the caller's end, because callers don't care whether the queue exists or needs to be created.

Reply:
Sorry, it took me a little while to understand the code due to the bad naming of the variables.  You are right in pointing out that default values are not considered here.  I will refactor to fix this.  I don't want to split up into `new` and `get` because that'd duplicate the try-to-get/create-if-if-it-does-not-exist dance in each caller.

Reply:
We will need to remove the consistency check for consumer parameters anyway, because it will produce misleading messages when a plugin configuration changes.

Reply:
https://konghq.atlassian.net/browse/KAG-698